hash.c

```diff
@@ -2936,19 +2982,44 @@ NOINSERT_UPDATE_CALLBACK(hash_aset_str)
 VALUE
 rb_hash_aset(VALUE hash, VALUE key, VALUE val)
 {
-    bool iter_p = hash_iterating_p(hash);
-
     rb_hash_modify(hash);
 
     if (RHASH_TYPE(hash) == &identhash || rb_obj_class(key) != rb_cString) {
-        RHASH_UPDATE_ITER(hash, iter_p, key, hash_aset, val);
+        RHASH_UPDATE(hash, key, hash_aset, val);
     }
     else {
-        RHASH_UPDATE_ITER(hash, iter_p, key, hash_aset_str, val);
+        RHASH_UPDATE(hash, key, hash_aset_str, val);
     }
+
     return val;
 }
 
+VALUE
+rb_hash_update_value(VALUE hash, VALUE key, VALUE val)
+{
+    bool iter_p = hash_iterating_p(hash);
+
+    rb_hash_modify(hash);
+
+    st_update_callback_func *func = (RHASH_TYPE(hash) == &identhash || rb_obj_class(key) != rb_cString)
+        ? UPDATE_CALLBACK(iter_p, hash_update_value)
+        : UPDATE_CALLBACK(iter_p, hash_update_value_str);
+
+    struct update_arg update_info = {
+        .arg = val,
+        .func = func,
+        .hash = hash,
+        .key  = key,
+        .value = val,
+    };
+
+    tbl_update2(&update_info);
+
+    const VALUE pre_existing_value = update_info.arg;
+
+    return pre_existing_value != Qundef ? pre_existing_value : Qnil;
```

Comment:
```suggestion
    rb_hash_modify(hash);
    RHASH_UPDATE(hash, key, hash_exchange_value, val);
    return update_info.arg;
```

Reply:
I don't think this works, because you don't have access to the `update_info.arg` that you're trying to return.

Reply:
```C
static int
hash_exchange_value(st_data_t *key, st_data_t *val, struct update_arg *arg, int existing)
{
    st_data_t *arg_value = (st_data_t *)arg->arg;
    st_data_t pre_existing_value = existing ? *val : Qnil;
    *val = *arg_value;
    *arg_value = pre_existing_value;

    return ST_CONTINUE;
}

NOINSERT_UPDATE_CALLBACK(hash_exchange_value)

static VALUE
rb_hash_exchange_value(VALUE hash, VALUE key, VALUE val)
{
    st_data_t value = (st_data_t)val;
    rb_hash_modify(hash);

    RHASH_UPDATE(hash, key, hash_exchange_value, &value);
    return (VALUE)value;
}
```

Reply:
Neat! This is great.